packages/marshal/src/encodeToSmallcaps.js

```diff
@@ -292,177 +296,138 @@ export const makeDecodeFromSmallcaps = ({
    * @param {SmallcapsEncoding} encoding must be hardened
    */
   const decodeFromSmallcaps = encoding => {
-    if (typeof encoding === 'string') {
-      switch (encoding.charAt(0)) {
-        case "'": {
-          // un-hilbert-ify the string
-          return encoding.slice(1);
-        }
-        case '@': {
-          return passableSymbolForName(encoding.slice(1));
+    switch (typeof encoding) {
+      case 'boolean':
+      case 'number': {
+        return encoding;
+      }
+      case 'string': {
+        if (encoding === '') {
+          return encoding;
         }
-        case '#': {
-          switch (encoding) {
-            case '#undefined': {
-              return undefined;
-            }
-            case '#NaN': {
-              return NaN;
-            }
-            case '#Infinity': {
-              return Infinity;
+        switch (encoding[0]) {
+          case "'": {
+            // un-hilbert-ify the string
+            return encoding.slice(1);
+          }
+          case '%': {
+            return passableSymbolForName(encoding.slice(1));
+          }
+          case '#': {
+            switch (encoding) {
+              case '#undefined': {
+                return undefined;
+              }
+              case '#NaN': {
+                return NaN;
+              }
+              case '#Infinity': {
+                return Infinity;
+              }
+              case '#-Infinity': {
+                return -Infinity;
+              }
+              default: {
+                assert.fail(X`unknown constant "${q(encoding)}"`, TypeError);
+              }
             }
-            case '#-Infinity': {
-              return -Infinity;
+          }
+          case '+':
+          case '-': {
+            const last = encoding.length - 1;
+            assert(
+              encoding[last] === 'n',
+              X`Encoded bigint must start with "+" or "-" and end with "n": ${encoding}`,
+            );
+            return BigInt(encoding.slice(0, last));
+          }
+          case '$': {
+            const result = decodeRemotableFromSmallcaps(encoding);
+            if (passStyleOf(result) !== 'remotable') {
+              assert.fail(
+                X`internal: decodeRemotableFromSmallcaps option must return a remotable: ${result}`,
+              );
             }
-            default: {
-              assert.fail(X`unknown constant "${q(encoding)}"`, TypeError);
+            return result;
+          }
+          case '&': {
+            const result = decodePromiseFromSmallcaps(encoding);
+            if (passStyleOf(result) !== 'promise') {
+              assert.fail(
+                X`internal: decodePromiseFromSmallcaps option must return a promise: ${result}`,
+              );
             }
+            return result;
+          }
+          default: {
+            return encoding;
```

Comment:
Should we enforce that the string doesn't start with one of the other currently unused hilberted chars? Maybe we can add `case` statements for those ?

Reply:
Done